Adding related items to plannings and events

tomaskikutis opened the pull request:
STT-67
STT-8

## Front-end checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] This pull request is adding missing TypeScript types to modified code segments where it's easy to do so with confidence
- [x] This pull request is using TypeScript interfaces instead of prop-types and updates usages where it's quick to do so
- [x] This pull request is using `memo` or `PureComponent` to define new React components (and updates existing usages in modified code segments)
- [x] This pull request is replacing `lodash.get` with optional chaining and nullish coalescing for modified code segments
- [x] This pull request is not importing anything from client-core directly (use `superdeskApi`)
- [x] This pull request is importing UI components from `superdesk-ui-framework` and `superdeskApi` when possible instead of using ones defined in this repository.
- [x] This pull request is not using `planningApi` where it is possible to use `superdeskApi`
- [x] This pull request is not adding redux based modals
- [x] In this pull request, properties of redux state are not being passed as props to components; instead, we connect it to the component that needs them. Except components where using a react key is required - do not connect those due to performance reasons.
- [x] This pull request is not adding redux actions that do not modify state (e.g. only calling angular services; those should be moved to `planningApi`)

tomaskikutis committed:
drag event for events and planning items

tomaskikutis committed:
drag and drop planning items to "related plannings" field of an event
support removing items from "related plannings" field.

tomaskikutis committed:
drag and drop related events to planning items

tomaskikutis committed:
show a message if item is already added

tomaskikutis committed:
improve code to use onChange

tomaskikutis committed:
make focusing an item after scrolling into view optional
before it was expanding an associated event which should not happen

tomaskikutis committed:
expose getting item action

tomaskikutis committed:
allow adding related events to open planning item from list view 3-dot menu

tomaskikutis committed:
add events as related from bulk actions

tomaskikutis committed:
add an option to disable focusing after adding a planning item to the editor

tomaskikutis committed:
add related planning items to event editor

tomaskikutis committed:
Merge remote-tracking branch 'upstream/feature/multiple-events-in-planning' into dragging-events-and-planning-items-by-tomas

tomaskikutis committed:
fix unit tests

tomaskikutis committed:
fix accidental passing of event as an argument to addPlanningItem

tomaskikutis committed:
only auto-expand coverages for newly created items

tomaskikutis commented:
@MarkLark86 If an existing planning item gets drag and dropped as related to an event, I'm adding it using the secondary link. Should it be possible to add coverages to planning items linked using the secondary link type?

tomaskikutis committed:
revert test config

MarkLark86 commented on client/actions/events/api.ts:
```diff
@@ -547,6 +549,58 @@ const uploadFiles = (event) => (
     }
 );
 
+function getLinkedPlanningItems(eventId: string): Promise<IRestApiResponse<IPlanningItem>> {
```
I'm thinking this function might be useful in other places, so it would be good to have this available under `IPlanningAPI.events.getLinkedPlanningItems` (to be implemented in [client/api/events.ts](https://github.com/superdesk/superdesk-planning/blob/develop/client/api/events.ts#L217))

MarkLark86 replied:
Would it be good to just return `Array<IPlanningItem>` instead of `IRestApiResponse<IPlanningItem>`? The function name states `get me the list of Planning items`, or do you see the need for attributes under the `IRestApiResponse` as well?

MarkLark86 commented on client/actions/events/api.ts:
```diff
@@ -547,6 +549,58 @@ const uploadFiles = (event) => (
+        const currentlyLinked: Array<IPlanningItem> = res._items;
+
+        const currentLinkedIds = new Set(currentlyLinked.map((item) => item._id));
+        const toLink: Array<IPlanningItem['_id']> = planningIds.filter((id) => currentLinkedIds.has(id) !== true);
```
[Set.has](https://tc39.es/ecma262/multipage/keyed-collections.html#sec-set.prototype.has) and [Array.includes](https://tc39.es/ecma262/multipage/indexed-collections.html#sec-array.prototype.includes) already return a boolean value, what is the use case for explicitly checking for `!== true`?

tomaskikutis replied:
I find it quicker to understand code when negation comes after the condition, not before.

MarkLark86 replied:
Fair enough

MarkLark86 commented on client/components/Events/EventMetadata/index.tsx:
```diff
@@ -92,19 +96,51 @@ class EventMetadataComponent extends React.PureComponent<IProps> {
+                    <Spacer h gap="4">
+                        {
+                            showEditButton && (
+                                <button
```
Could it be possible to use a button from the UI-Framework instead?

MarkLark86 commented on client/components/fields/editor/AssociatedEvent.tsx:
```diff
@@ -15,28 +16,121 @@ interface IProps extends IEditorFieldProps {
+                ...this.getCurrentValue(),
+                {
+                    _id: id,
+                    link_type: 'secondary',
```
The `link_type` here could be different values, depending on the instance config.

We have a `PLANNING_EVENT_LINK_METHOD` config available in settings.py. This can be exposed to front-end by adding something like the following in [planning/__init__.py](https://github.com/superdesk/superdesk-planning/blob/feature/multiple-events-in-planning/server/planning/__init__.py#L241)

```python
app.client_config["planning"]["event_link_method"] = get_planning_event_link_method()
```

And adding `event_link_method` to [IPlanningConfig](https://github.com/superdesk/superdesk-planning/blob/develop/client/interfaces.ts#L273) and in [client/config.ts](https://github.com/superdesk/superdesk-planning/blob/develop/client/config.ts).

Then based on that config value:

**one_primary**
* Only 1 Event can be linked, therefor if an Event is already linked this action should not be available, and an error thrown here
* Otherwise the `link_type` would be `primary`

**many_secondary**
* `link_type` is always `secondary`

**one_primary_many_secondary**
* `link_type` is `primary` if there are no Event's currently linked
* `link_type` is `secondary` otherwise

MarkLark86 commented on client/utils/events.ts:
```diff
@@ -461,14 +464,73 @@ function canMarkEventAsComplete(
+ * Will return true if there is at least one event that can be added
+ */
+export function canAddSomeEventsAsRelatedToPlanningEditor(eventIds: Array<IEventItem['_id']>): boolean {
+    const editor = planningApi.editor(EDITOR_TYPE.INLINE);
```
This is where we would use that `event_link_method` config to add additional clause if this action should be available (i.e. config is `one_primary` and currentRelatedEvents is not empty, then return false)

MarkLark86 commented on client/utils/events.ts:
```diff
@@ -461,14 +464,73 @@ function canMarkEventAsComplete(
+            .filter((eventId) => isEventAlreadyAddedAsRelated(eventId) !== true)
+            .map((eventId): IPlanningRelatedEventLink => ({
+                _id: eventId,
+                link_type: 'secondary',
```
And same here, `link_type` would depend on the config and currently linked Events (if any)

MarkLark86 reviewed (commented):
Looks good, mostly just suggestion about implementing logic around the link config in the front-end

MarkLark86 commented:
> @MarkLark86 If an existing planning item gets drag and dropped as related to an event, I'm adding it using the secondary link. Should it be possible to add coverages to planning items linked using the secondary link type?

I might need a little more context here, as I'm not sure I understand the scenario entirely. When adding a Coverage to a Planning item, there is no logic involved around any linked Events that alllows or denies adding these Coverages. Am I missing something here, could you maybe explain more please

tomaskikutis committed:
fix lint

tomaskikutis committed:
move getLinkedPlanningItems into events api

tomaskikutis committed:
use icon button from ui-framework

tomaskikutis committed:
expose planning event link method config to front-end

tomaskikutis committed:
extend global configs with planning specific options

tomaskikutis committed:
add planning_event_link_method to app config interface

tomaskikutis committed:
perform validation and display errors when adding related events to a planning item

tomaskikutis committed:
add related plannings to event editor

tomaskikutis committed:
set correct link type when adding related planning items to event

tomaskikutis committed:
pass full event

tomaskikutis committed:
let promise reject so .thens do not execute

tomaskikutis committed:
avoid circular import

tomaskikutis committed:
Merge remote-tracking branch 'upstream/feature/multiple-events-in-planning' into adding-related-items-to-plannings-and-events

tomaskikutis commented:
FYI I've force-pushed to revert a bad merge. I had by mistake merged develop to this branch, instead of `feature/multiple-events-in-planning`

tomaskikutis committed:
fix lint

tomaskikutis committed:
do not attempt to change links for items with temp IDs

tomaskikutis committed:
add brief instructions on how to run e2e tests locally

tomaskikutis commented:
> I might need a little more context here, as I'm not sure I understand the scenario entirely. When adding a Coverage to a Planning item, there is no logic involved around any linked Events that alllows or denies adding these Coverages. Am I missing something here, could you maybe explain more please

That's what I wanted to know in the end - whether coverages have some relation to the planning item and its link type. [Because in the UI](https://github.com/user-attachments/assets/9a4c8016-162a-49ed-a350-3ef7f06fe0ca) they look attached to planning items. Since you say they're not connected - all good then - I don't have to do anything special to address it.

tomaskikutis commented on client/actions/events/api.ts:
```diff
@@ -547,6 +549,89 @@ const uploadFiles = (event) => (
+                const ageSeconds = (now.getTime() - createdAt.getTime()) / 1000;
+                const tooRecent = ageSeconds > 30;
+
+                if (tooRecent) {
```
@MarkLark86 any idea how to avoid this hack?

MarkLark86 replied:
Only thing that comes to mind is if you also pass in the original associatedPlanning IDs. Check non-temp IDs (aka not "fake ID") that are in original that are not in updates are to be unlinked.

tomaskikutis committed:
fix lint

tomaskikutis committed:
fix comparison symbol

tomaskikutis committed:
Merge branch 'feature/multiple-events-in-planning' into adding-related-items-to-plannings-and-events

tomaskikutis committed:
improve docs on e2e

tomaskikutis committed:
Merge branch 'feature/multiple-events-in-planning' into adding-related-items-to-plannings-and-events

tomaskikutis committed:
address circular dependency issue
actions.planning.api was evaluating to undefined

tomaskikutis committed:
use icon button for linking events/plannings

petrjasek committed:
fix server test

tomaskikutis commented:
@MarkLark86 besides the hack, anything else you'd like me to address here?

MarkLark86 commented on client/actions/events/api.ts:
```diff
@@ -574,9 +660,22 @@ const save = (original, updates) => (
+                    return Promise.resolve([updatedEvent]);
+                }
+
+                // return Promise.resolve([updatedEvent]);
```
Remove commented code

MarkLark86 commented on client/globals.d.ts:
```diff
@@ -147,3 +147,44 @@ declare var ResizeObserverSize: {
+
+// KEEP IN SYNC WITH client/planning-extension/src/globals.d.ts
+declare module 'superdesk-api' {
+    interface ISuperdeskGlobalConfig {
```
Nice one :+1: Removes the need for the hack `const appConfig = config.appConfig as IPlanningConfig;`.
Though, where is it getting the references to types from client/interfaces.ts (such as IPlanningItem)?

tomaskikutis replied:
Good catch. I forgot to import it so it wasn't getting references and evaluating to `any`

MarkLark86 commented on server/planning/__init__.py:
```diff
@@ -242,6 +242,9 @@ def init_app(app):
     ] = get_config_default_create_planning_series_with_event_series(app)
 
+    # TODO: use get_planning_event_link_method
+    app.client_config["planning_event_link_method"] = "one_primary_many_secondary"
```
TODO action needs to be fixed

tomaskikutis replied:
@petrjasek can you fix this one for me?

MarkLark86 reviewed (approved):
@tomaskikutis Looks good to me.

tomaskikutis committed:
remove commented out code

tomaskikutis committed:
add missing imports

petrjasek committed:
fix server TODO

tomaskikutis committed:
fix related items not working in a modal editor
fix a mistake where inline editor was always used to get a editor API instance instead of passing the type of currently open editor

petrjasek committed:
add notifications when links are added/removed

petrjasek committed:
add `links` to `event:link_updated` event

tomaskikutis committed:
improve how related plannings are kept up to date for events
it wasn't updating when related plannings were being removed

tomaskikutis committed:
don't use an expression where return value isn't used

tomaskikutis committed:
fix a locking-related bug

when a related event gets removed from a planning item and saved - planning editor gets into a read-only mode.

The cause was that when opening an editor with a planning item having related events, only event would get into "locks" section in redux store. Planning item would't. If no related events are present, planning item would get to the "locks" section.

This commit changes the behavior for planning items with related events so both the planning item itself and the related event get into the "locks" section.

tomaskikutis committed:
WIP on settling locks after saving a planning item

tomaskikutis committed:
recompute soft locks on planning update

tomaskikutis committed:
add z-index so it is visible in modal based editor

tomaskikutis committed:
fix unit tests

tomaskikutis commented:
@petrjasek could you take a look why `CI-Server / server-behave` started crashing?

MarkLark86 commented:
> @petrjasek could you take a look why `CI-Server / server-behave` started crashing?

@tomaskikutis Ketan fixed this in a [PR](https://github.com/superdesk/superdesk-planning/pull/2138) for develop branch.

tomaskikutis commented:
@petrjasek could you merge develop to `feature/multiple-events-in-planning`? There are back-end conflicts there

petrjasek commented:
@MarkLark86 can you have a look? seems a bit complicated for me

tomaskikutis committed:
Merge branch 'feature/multiple-events-in-planning' into adding-related-items-to-plannings-and-events
